### PR TITLE
Change the warning statement for unused result from expression

### PR DESCRIPTION
**Pull request by fluffybeing**
Resolves [SR-5983](https://bugs.swift.org/browse/SR-5983).

### CODE REVIEW

**Comment by huonw** [comment]
Thanks!

The compiler also has tests for diagnostics like this, using directives like `expected-warning {{...}}` and `expected-error {{...}}` to tell the compiler that it should find an specific message on the line with the comment (or, with `expected-warning @number` on the line `number` before or after). There's a few tests that look for this one, could you update them too?

```
test/attr/attr_discardableResult.swift
152:  a() // expected-warning {{result of call is unused, but produces 'Int'}}

test/ClangImporter/objc_parse.swift
217:  // expected-warning @-1 {{result of call is unused, but produces 'Unmanaged<AnyObject>!'}}

test/Constraints/diagnostics.swift
340:f8(10)          // expected-warning {{result of call is unused, but produces 'Int'}}

test/expr/expressions.swift
674:  fn(&n, 12) // expected-warning {{result of call is unused, but produces 'Int'}}

test/Parse/recovery.swift
131:  } while { true }() // expected-error{{missing condition in a 'while' statement}} expected-error{{consecutive statements on a line must be separated by ';'}} {{10-10=;}} expected-warning {{result of call is unused, but produces 'Bool'}}
```

File: include/swift/AST/DiagnosticsSema.def [hunk 1]
```diff
@@ -2814,7 +2814,7 @@ WARNING(expression_unused_result_call,none,
 WARNING(expression_unused_result_operator,none,
         "result of operator %0 is unused", (DeclName))
 WARNING(expression_unused_result_unknown, none,
-        "result of call is unused, but produces %0", (Type))
+        "result of call to function returning %0 is unused", (Type))
```
Comment (CodaFi) [inline include/swift/AST/DiagnosticsSema.def]: Nit: You may not necessarily be calling a function if this is emitted, you may be calling a closure.

I think you can probably delete this and just use `expression_unused_result_call` if it's all the same.

Comment (rahulrrixe) [reply include/swift/AST/DiagnosticsSema.def]: @CodaFi I will update it.

**Comment by rahulrrixe** [comment]
@huonw thanks for the help. Also, I was curious that how can I run the test locally. Wasn't able to figure out from the documentation.

**Comment by rahulrrixe** [comment]
@huonw am I missing something? or this PR is ready for acceptance.

File: include/swift/AST/DiagnosticsSema.def [hunk 2]
```diff
@@ -2814,7 +2814,7 @@ WARNING(expression_unused_result_call,none,
 WARNING(expression_unused_result_operator,none,
         "result of operator %0 is unused", (DeclName))
 WARNING(expression_unused_result_unknown, none,
-        "result of call is unused, but produces %0", (Type))
+        "result of call to closure returning %0 is unused", (Type))
```
Comment (CodaFi) [inline include/swift/AST/DiagnosticsSema.def]: This is not quite what I was pointing out - not everything this diagnoses is a call to a closure, nor is everything it diagnoses a call to a function.  So you have two options:

1) Conditionalize this diagnostic on whether the called value is a closure or a function.
2) Just delete it and use `expression_unused_result_call` instead.

I really recommend #2.  This diagnostic seems genuinely superfluous.

Comment (rahulrrixe) [reply include/swift/AST/DiagnosticsSema.def]: @CodaFi Thanks for the context. I thought it is just for closure and will not prefer conditionalizing it. I will update it then.

**Comment by CodaFi** [comment]
This was addressed by #16059.  Thanks!

**Pull request closed without merging**